### ISSUE TITLE
With Pandoc 2.8, attributes are added to the header element instead of its parent when --section-divs is enabled

### ISSUE THREAD

**Issue by yihui** [jgm/pandoc#5965]
I noticed a change in Pandoc 2.8 (which might be from 9f984ff26ac248a27212a37ab34754a2e9261e8c), but I'm not sure it is intended. Before 2.8, attributes of a header are added to its parent `div` (or `section`) when `--section-divs` is enabled, such as the class attribute:

```sh
$ pandoc -t html4 --section-divs <<< "# Hello {.foo}" 

<div id="hello" class="section level1 foo">
<h1>Hello</h1>
</div>
```

With 2.8+, the class attribute is added to `h1`:

```sh
$ pandoc -t html4 --section-divs <<< "# Hello {.foo}" 

<div id="hello" class="section level1">
<h1 class="foo">Hello</h1>
</div>
```

(Note that the `id` attribute is still on the `div` instead of `h1`, though.)

We have plenty of ([R Markdown](https://github.com/rstudio/rmarkdown)) applications that depend on the previous behavior, i.e., the `class` attribute being added to the parent `div` instead of the header itself. For example, previously we could style a whole section by defining CSS rules for `.foo { }`. Now these rules are only applied to the header elements. We also have JavaScript applications that rely on this behavior. I wonder if there is a chance to restore the old behavior. Thank you!

**Comment by jgm** [comment]
Yes, it was intentional.  But maybe not a good idea, if it's going to break lots of things.  It's probably not too late to role this back.

**Comment by jgm** [comment]
I'm about to push a change which would yield the following:
```html
<div id="hello" class="section level1 foo">
<h1 class="foo">Hello</h1>
</div>
```
That would be okay?

**Comment by yihui** [comment]
That will help a lot to us! I still wish the class is only added to the `div` and not repeated on `h1`, because previously we could use the selector `.foo` in CSS/JavaScript without specifying the HTML element, and now our selectors have to be changed to `div.foo` (and `section.foo` for HTML5). By comparison, without the class `foo` on `h1`, it is still easy to select the header (`.foo > h1`). Thank you!

**Comment by jgm** [comment]
I'd consider stripping the class and other attributes from the header. However, I think we need some of them to be on the header itself, e.g. the `number` attribute and the `unlisted` or `unnumbered` classes.  It seemed simpler to leave them all on, but maybe that's not the best approach?

**Comment by yihui** [comment]
How about providing a flag like `--section-attributes=section|header|both` so users can decide by themselves? To me, I want the attributes to be on the `div` only, so I'd choose `section`, but you could make `both` as the default.

**Comment by jgm** [comment]
Putting all attributes on `section` alone would break certain things (e.g. current behavior of `unnumbered` or `unlisted` for LaTeX output).

I would really rather avoid multiplying command line options unless there's a really good reason.

Seems like it's just a slight inconvenience for you to have them on headers too, right?

**Comment by yihui** [comment]
Okay. If it is too much trouble for you, please feel free to consider this issue resolved. I guess I can deal with the redundant classes on the headers. Again, thank you very much!

**Comment by jgm** [comment]
I'll keep this open a bit longer to think about this issue.

**Comment by tarleb** [comment]
I wonder whether could make sense to handle `--section-divs` as a "transformation", as that would allow to modify the intermediate document with a filter. Adding/removing classes to headers and divs would be much easier that way.

**Comment by jgm** [comment]
Currently the "transformations" get applied before the filters, though.

**Comment by tarleb** [comment]
My thought was that this way, filters could modify the doc after transformation has "hierarchicalized" the AST. On the other hand, if one wants to do that, it might make more sense to do *everything* in the filter.

E.g., this restores the original behavior (if one replaces `--section-divs` with `--lua-filter=section-divs.lua`:

``` lua
-- file: section-divs.lua
function hierarchicalize (doc)
  return pandoc.Pandoc(
    pandoc.utils.make_sections(true, nil, doc.blocks),
    doc.meta
  )
end

function add_classes (div)
  local header = div.content[1]
  if header and header.t == "Header" then
    div.classes:extend(header.classes)
    div.classes:extend({"level" .. header.attributes.number})
    header.classes = {}
    header.attributes.number = nil
    return div
  end
end

return {
  {Pandoc = hierarchicalize},
  {Div = add_classes}
}
```

**Comment by jgm** [comment]
Ah, I see.  It wouldn't be an easy change to make this a transformation.  Currently `makeSections` gets called in the writers (and in some shared functions), and for various reasons (not just the use of `--section-divs`).  Moving all of that logic to App may not make sense, but it may be worth exploring.

Or we could try to restore more or less the old behavior, where everything (except possibly things like `unlisted` if they need to be on headers) goes to the section Div.

**Comment by jgm** [comment]
Let's try this for now, where the attributes are the same on headings and divs.
It seems fairly harmless.

**Comment by gwern** [comment]
To reopen this: I would like to endorse yihui's request that the classes be put on the section rather than both section & header.

I too do section by section restyling on gwern.net (primarily my very useful 'collapsed sections' functionality), and we tracked down some issues in sections not expanding correctly to the redundant classes, which was a surprise. (Not least because no documentation mentions this, not even the relevant lines in the HTML writer when we dug down to the source code to figure out where these erroneous duplications came from. The `writerSectionDivs` documentation is out of date, incidentally.)

If it could be made div/section only, that'd be best; if not, an option would at least work. For now I guess we'll try to work around this in JS.

**Comment by jgm** [comment]
ok I'll reopen for further consideration

**Comment by mb21** [comment]
I guess you can always use the CSS selector `section.myClass`? same when using js with `document.querySelector`

**Comment by reagle** [comment]
In this [slide][1] you can see that "big text" is pushed off the right side. The [problem][2] is the 'r-fit-text' class attribute on the `section` element and its child `h2`. If I remove it from `section`, the big text works fine -- but not vice versa.

I'm not sure if this is a pandoc or revealjs issue. I just upgraded to revealjs 4.1.1, and I've followed newer  pandocs (now 2.14.0.1) since I last ran these tests.

[1]: https://reagle.org/joseph/talks/2014/exemplar.html#/big-text
[2]: https://pandoc.org/try/?text=%23%23+big+text+%7B.r-fit-text%7D%0A&from=markdown&to=revealjs&standalone=0

**Comment by reagle** [comment]
Tis issue hasn't been touched since the summer and while it's still in the "next release" milestone, there's been a few releases since then. If that is the case, I assume that the double-decorations are a done deal? If so, any suggestions on how I can undo their effect on `r-fit-text` within revealjs?

**Comment by jgm** [comment]
Forgot about this one.
Of course, making changes here is a bit risky as it may break things...

**Comment by reagle** [comment]
Is there a possibility of tweaking pandoc's output specifically for the revealjs format to resolve this and two other issues? (I don't think these can be accomplished via the [default.revealjs template](https://github.com/jgm/pandoc-templates/blob/master/default.revealjs).

There's three things I'd like to use which are broken/incompatible between the structure pandoc creates and the structure revealjs expects.

1. This issue, #5965: redundant `section/h2` misplaces "big text" as [seen here](https://reagle.org/joseph/talks/2014/exemplar.html#/big-text).
2. a single `r-stretch` element must a child of the `section`; so `img` can’t be within `p, frame` or `a`, as [seen here](https://reagle.org/joseph/talks/2014/exemplar.html#/local-html-r-stretch).
3. when image is in link, pandoc does not create figure or figcaption, as [seen here](https://reagle.org/joseph/talks/2014/exemplar.html#/caption-link). Perhaps related to #3928.

**Comment by jgm** [comment]
@reagle I've just a fixed for the r-fit-text issue.

**Comment by jgm** [comment]
Looks like we already had code to address "stretch" on images, but the class name was changed from "stretch" to "r-stretch". I'll adjust this.

**Comment by jgm** [comment]
> when image is in link, pandoc does not create figure or figcaption,

True, but that's because the implicit_figures extension says that the image must be the only thing in the paragraph.
You can always put a link on the caption text.

**Comment by reagle** [comment]
Excellent! I'll be sure to try the two fixes in the next release. And thanks for the link-caption-instead suggestion.

BTW: Did you change the [mention of `stretch` to `r-stretch`](https://pandoc.org/MANUAL.html#extension-implicit_figures) class in the README as well?

**Comment by jgm** [comment]
Thanks, I'll update the manual too.

**Comment by reagle** [comment]
Thanks, this fixed the issue for revealjs. (Should this issue remain open then?)

**Comment by ricopicone** [comment]
> Putting all attributes on `section` alone would break certain things (e.g. current behavior of `unnumbered` or `unlisted` for LaTeX output).
> 
> I would really rather avoid multiplying command line options unless there's a really good reason.
> 
> Seems like it's just a slight inconvenience for you to have them on headers too, right?

Here's a use-case I ran into that makes the current double-placement of Header classes annoying and the option suggested nice. In js I hide/unhide sections by adding/removing a `hide` class. I would like to set the default as hidden and unhide/hide with js. I can't do this now because the Headers would also get the `hide` class, and I'd have to process these in js as well, which kind of defeats the run-time performance gains.

In some ways, the trouble is that we can't just set classes on the `section`s because they're not directly accessible via a Lua filter (afaict). I might solve this by re-`pandoc`ing the output so I can access them, but I'm anticipating some lossiness I'll have to fix.

Anyway, this is a tricky problem and I appreciate that it's being considered. I do think the solution suggested by @yihui of the suboptions `--section-attributes=section|header|both` would work for me.

**Comment by jgm** [comment]
> the trouble is that we can't just set classes on the sections because they're not directly accessible via a Lua filter (afaict)

You can add the sections in Lua using the approach demonstrated by @tarleb in his comment, above.  Then they'll be directly accessible.  It's a bit clumsy but possible.

**Comment by ricopicone** [comment]
Thanks @jgm -- didn't see that before. It works for me!
